web-gateway/app/controllers/TransactionController.java

```diff
@@ -0,0 +1,147 @@
+                                    winner = Optional.of(u);
+                                }
+                            }
+                            Currency currency = Currency.valueOf(transaction.getItemData().getCurrencyId());
+                            return ok(views.html.transaction.render(showInlineInstruction, Optional.of(transaction), seller, winner, Optional.of(currency), Optional.empty(), nav));
+                        }
+                        else {
+                            String msg = ((TransportException) exception.getCause()).exceptionMessage().detail();
+                            return ok(views.html.transaction.render(showInlineInstruction, Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.of(msg), nav));
+                        }
+                    });
+                })
+        );
+    }
+
+    public CompletionStage<Result> submitDeliveryDetailsForm(String id) {
+        return requireUser(ctx(), user ->
+            loadNav(user).thenCompose(nav -> {
+                UUID itemId = UUID.fromString(id);
+                CompletionStage<TransactionInfo> transactionFuture = transactionService.getTransaction(itemId).handleRequestHeader(authenticate(user)).invoke();
+                return transactionFuture.handle((transaction, exception) ->{
+                    if(exception == null) {
+                        DeliveryDetailsForm form = new DeliveryDetailsForm();
+                        Optional<DeliveryInfo> maybeDeliveryInfo = transaction.getDeliveryInfo();
+                        if(maybeDeliveryInfo.isPresent()) {
+                            form.setAddressLine1(maybeDeliveryInfo.get().getAddressLine1());
+                            form.setAddressLine2(maybeDeliveryInfo.get().getAddressLine2());
+                            form.setCity(maybeDeliveryInfo.get().getCity());
+                            form.setState(maybeDeliveryInfo.get().getState());
+                            form.setPostalCode(maybeDeliveryInfo.get().getPostalCode());
+                            form.setCountry(maybeDeliveryInfo.get().getCountry());
+                        }
+                        return ok(
+                                views.html.deliveryDetails.render(
+                                        showInlineInstruction,
+                                        !transaction.getCreator().equals(user),
+                                        itemId,
+                                        formFactory.form(DeliveryDetailsForm.class).fill(form),
+                                        transaction.getStatus(),
+                                        Optional.empty(),
+                                        nav)
+                        );
+                    }
+                    else {
+                        String msg = ((TransportException) exception.getCause()).exceptionMessage().detail();
```

Comment:
I'm not sure that it's safe to always assume that `exception.getCause()` will always be a `TransportException`.